### PR TITLE
[BugFix] Add multiref when grad-reduce/no-grad-reduce are mixed used

### LINKED ISSUE #31: Refine multiref implementation for PR #29 (no-grad-reduce annotation)
[#29 ](https://github.com/msrasys/nnscaler/pull/29)  introduces a “no-grad-reduce” annotation mechanism for custom op shape annotations so that, for specific partition identifiers, nnscaler can skip inserting gradient all-reduce adapters (avoiding incorrect or redundant reductions).

But this leads a bug for multiref. The current multiref logic only looks at the partitioning of the tensor, without considering the partitioning of tensor.grad. After adding this feature, it needs to look at the partitioning of both the tensor and tensor.grad simultaneously.

Repro:

```
@nnscaler.register_op('a b : /, b c -> a c') #<- this will fail
#@nnscaler.register_op('a b, b c -> a c') #<- this can work
def _shared_skip_op(x, w):
    """Custom op that asks nnscaler to skip grad all-reduce on input 0."""
    return x @ w


class _SharedInputModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.w0 = torch.nn.Parameter(torch.randn(4, 4))
        self.w1 = torch.nn.Parameter(torch.randn(4, 4))
        self.w2 = torch.nn.Parameter(torch.randn(4, 4))

    def forward(self, raw_x):
        # The same `x` is consumed by two operators:
        #   - `_shared_skip_op`: annotation declares input 0 = ': /'
        #     so its grad-reduce should be skipped.
        #   - `torch.matmul`: standard op, full grad-reduce expected.
        x = torch.nn.functional.linear(raw_x, self.w0)
        a = _shared_skip_op(x, self.w1)
        b = torch.matmul(x, self.w2)
        return (a + b).sum()


def _shared_input_partition_policy(graph, cfg):
    """Partition both consumers along weight-dim 1 (= identifier 'c'),
    which causes `x` (replicated input) to need grad reduction on the
    `torch.matmul` side. The `_shared_skip_op` side is annotated with `: /`,
    so its grad reduction must be skipped — but this MUST NOT alter the
    `torch.matmul` side's gradient bookkeeping.
    """
    from nnscaler.policies import OpPlan, OpPartition, get_pas_ops

    for node in get_pas_ops(graph):
        if node.fn == torch.matmul:
            yield OpPlan(node, partition=OpPartition(input=1, dim=1))
        elif node.fn == _shared_skip_op:
            yield OpPlan(node, partition=OpPartition(input=1, dim=1))


@replace_all_device_with('cpu')
def test_shared_input_with_no_grad_reduce_consumer(tmp_path):
    """Two consumers of an internal activation `x`:
      * `_shared_skip_op` is annotated with `: /` -> grad-reduce skipped;
      * `torch.matmul` is a standard op -> grad-reduce required.

    Both are partitioned along weight-dim 1 (identifier `c`), so `x` is
    replicated across both ranks for both consumers. After fix, parallelize
    should succeed and emit at least one `identity_allreduce` on the matmul
    side. Currently `parallelize` itself raises before any code is generated.
    """
    m = _SharedInputModel()
    m.train()
    parallelize(
        m,
        {'raw_x': torch.randn(4, 4)},
        _shared_input_partition_policy,
        ComputeConfig(2, 4, use_end2end=True),
        gen_savedir=tmp_path,
        load_module=False,
        reuse='override',
    )

    from pathlib import Path
    gencode_text = ''.join(
        f.read_text() for f in Path(tmp_path).glob('**/gencode*.py')
    )
    n_identity_allreduce = gencode_text.count('identity_allreduce')
    assert n_identity_allreduce > 0, (
        "expected at least one identity_allreduce (on torch.matmul's x input); "
        f"got {n_identity_allreduce}."
    )
 
```

### PR DESCRIPTION
**Pull request by 0xWJ**
This pull request improves the handling of replicated tensors and gradient reduction logic in `fn` policy. It introduces a more precise distinction between replicated tensors that require gradient all-reduce and those that do not, so we can add `multiref` more precisely.

### CODE REVIEW

File: nnscaler/policies.py
Comment (zyeric) [inline nnscaler/policies.py:838]: reducer generation may crash

File: nnscaler/policies.py [hunk 1]
```diff
@@ -804,7 +818,8 @@ def fn(
             stage = pp_segs[idx]
             split_list = list(splits)
             if len(split_list) > 1 or (
-                is_seg_output[idx] and split_list[0] is not None # treat segment output as a consumer
+                # treat segment output as a consumer (replicated but not all-reduced)
+                is_seg_output[idx] and split_list[0] != 'rn'
```
Comment (zyeric) [inline nnscaler/policies.py]: may be related to the `loss` tensor

**Comment by 0xWJ** [comment]
@zyeric  Thank you for your clarification. I have found the original PR in devops, and have better understanding about the problem. 

I have added 3 possible solutions for the loss problem. Please help to check the test cases.

Solution 1: Never allow the case when segment output shares the same tensor with input of other ops. This is done in `fn` by inserting Identity ops.
Solution 2: The user can manually insert an Identity op.
Solution 3: The user can manually insert a MultiRef op.

I also added a test case for shared parameters in PP. Please also check the test case.

Thank you very much !

**Review by lynex — APPROVED** [review]

**Review by lynex — APPROVED** [review]
